mongo-shell-basic-mongo-queries-review (#9)
mattdavis0351/mongodb-labs: exercises/mongo-shell/basic-mongo-queries.md

# CRUD Operations

> This document deals with basic create, read, update and delete operations using the mongo shell. 

Reviewer: ```suggestion
This document deals with basic **Create**, **Read**, **Update** and **Delete** operations using the mongo shell. 

```

I think distinction that these are the 'types' of operations is helpful when you are new to acronyms.

[Read more](https://docs.mongodb.com/manual/crud/) :book: about the CRUD operations here. 

## Create Operations

> A MongoDB database is made up of 1 or more collections which are in turn made up of 1 or more documents. A document has a `key: value` structure, similar to JSON objects. In the following exercises, we'll learn how to insert documents using the following operations:

Reviewer: ```suggestion
A MongoDB database is made up of 1 or more collections which are in turn made up of 1 or more documents. A document has a `key: value` structure, similar to JSON objects. In the following exercises, we'll learn how to insert documents using the following methods:

```

* `insertOne()`
* `insertMany()` 

**Exercise 1** :computer: 

* Verify what database you are currently using (should be `test`, this is the default.)

> 

Reviewer: ```suggestion
MongoDB makes this a super easy step.  From the Mongo-Shell type `db` and hit `Enter`, the output is the current working database.

**query:**
```

    ```javascript
    db
    ```

* Let's create a new database called `employee`. 
    ```javascript
    use employee_db
    ```

:arrow_right: Note that `use` creates a new database, if one does not already exist. 

> * Next, let's create a collection called `employee_info`.

Reviewer: How do we do so?  How is this similar/different from SQL which they have been writing the entire time?

Author: look at the rest of the suggestions and let me know what you think

Reviewer: 🗑

Author: :fu:

    ```javascript
    db.createCollection("employee_info")
    ```
    
* Let's confirm that the collection was successfully created.

    ```javascript
    show collections
    ```

> :arrow_right: Similarly, you can also use `show dbs` to obtain a list of the all the databases. 

Reviewer: 👍

> * Let's try to insert the following `key:value` pairs as a document in our collection. 

Reviewer: ```suggestion
* Let's try to insert the following object, a set of `key:value` pairs surround by `{}`,  as a document in our collection. 
```

>     {

Reviewer: What about a table showing this information before the object?  This would allow them to see table data converted to JSON.  

Also a block of JSON is confusing here... what is this object, why does it exist.  We should preface whey they are seeing JSON right now.  Maybe something like "below you will find the data we are going to insert.  This is what it looks like as a table, but we need it as an object... oh hey, here's the object, okay let's insert it 👍 "

        "fname": "john",
        "lname": "doe",
        "salary": 70000,
        "departments": ["sales", "admin"],
        "hiredate": "2018-08-29"
      }

    **query:**
    ```javascript
    db.employee_info.insertOne({
        "fname": "john", 
        "lname": "doe", 
        "salary": 70000, 
        "departments": ["sales", "admin"], 
        "hiredate": "2018-08-29"})
    ```

    **result:**
    ```javascript
    {
        "acknowledged" : true,
        "insertedId" : ObjectId("5d8e6f9ccaa4f8ddbe27296f")
    }
    ```
    The response to this query is a document in itself as seen below. `"acknowldeged": true` means we the insert was successful. Also note the Object ID which is the unique identifier for this document. Every document has its own unique `_id`. It is not necessary to specify `_id` in the query in which case, MongoDB will automatically add one to the document. 

> 

Reviewer: ```suggestion
➡️ Letting Mongo take care of your `_id` field is a good idea.  Mongo will never try to assign a duplicate value to this field, sometimes we make mistakes and may try to do that.  Mongo is pretty relaxed and sometimes doesn't stop us from doing reckless things such as assigning the same `_id` to more than one Object.  If you need control over an `id` field for you document consider creating one instead of overwriting the object id.
```

**Exercise 2** :computer: 

Spend **5 minutes** :alarm_clock: to try inserting the following employees' information into the `employee_info` collection:

{
  "empno": 1,
  "fname": "charlie",
  "lname": "rodgers",
  "role": "manager"
},
{
  "empno": 2,
  "fname": "steve",
  "lname": "smith",
  "role": "team lead"
},
{
  "empno": 3,
  "fname": "sally",
  "lname": "jones",
  "role": "team lead"
},
{
  "empno": 4,
  "fname": "ben",
  "lname": "bradley",
  "role": "manager"
}

:arrow_right: You might be wondering whether the insert will work, given that some of the keys (field names) are different from the document we inserted earlier. And herein lies the beauty of MongoDB!!! Unlike relational databases, the same keys (field names) do not have to be present in all documents of the collection. :clap: 

> How many of you inserted the documents one by one? 

Reviewer: ```suggestion
:bulb: How many of you inserted the documents one by one? 
```

> It's not wrong but the more efficient way of inserting multiple documents is to use `insertMany()`. Simply replace `insertOne()` in the query above with `insertMany()`. ]

Reviewer: ```suggestion
It's not wrong but the more efficient way of inserting multiple documents is to use `insertMany()`. Simply replace `insertOne()` in the query above with `insertMany()`.
```

:arrow_right: Keep in mind that `insertMany()` takes an array as input so enlcose all your documents in `[]` as shown in the query below.

**query:**
```javascript
db.employee_info.insertMany([
    {
        "empno": 1,
        "fname": "charlie",
        "lname": "rodgers",
        "role": "manager" 
    }, 
    {
        "empno": 2,
        "fname": "steve",
        "lname": "smith",
        "role": "team lead" 
    },
    {
        "empno": 3,
        "fname": "sally",
        "lname": "jones",
        "role": "team lead" 
    }, 
    {
        "empno": 4,
        "fname": "ben",
        "lname": "bradley",
        "role": "manager" 
    }
])
```
**Exercise 3** :computer: 

You've got some more interesting employees! Spend **10 minutes** :alarm_clock: converting the following table to JSON and inserting it into our `employee_info` collection.

|name|title|interesting fact|address|
|:---:|:---:|---|---|

> |Harry Potter|Wizard|I kicked the shit out of Voldemort!|Hogwarts castle|

Reviewer: ```suggestion
|Harry Potter|Wizard|I kicked the sh*t out of Voldemort!|Hogwarts castle|
```

No bad words!!!

|Groot|Superhero|I can only say 3 words in this order: 'I am Groot'!|Planet X|
|Pikachu|Electric type pokemon|Want to see my ThuderBolt and Quick Attack?!| Pallet Town|

---

## Read Operations

> :warning: Before we dive into our read operations, make sure you can access the `movieDetails` collection. 

Reviewer: ``` suggestion
:warning: Before we dive into our **Read** operations, make sure you can access the `movieDetails` collection.

```

## Create Operations

A MongoDB database is made up of 1 or more collections which are in turn made up of 1 or more documents. A document has a `key: value` structure, similar to JSON objects. In the following exercises, we'll learn how to insert documents using the following methods:

* `insertOne()`
* `insertMany()` 

**Exercise 1** :computer: 

* Verify what database you are currently using (should be `test`, this is the default.)
MongoDB makes this a super easy step.  From the Mongo-Shell type `db` and hit `Enter`, the output is the current working database.

**query:**
    ```javascript
    db
    ```

* Let's create a new database called `employee_db`. 
    ```javascript
    use employee_db
    ```

:arrow_right: Note that `use` creates a new database, if one does not already exist. 

* Next, let's create a collection called `employee_info`.

    ```javascript
    db.createCollection("employee_info")
    ```
    
* Let's confirm that the collection was successfully created.

    ```javascript
    show collections
    ```
:arrow_right: Similarly, you can also use `show dbs` to obtain a list of the all the databases. 

> * Let's try to insert the following object, a set of `key:value` pairs surround by `{}`,  as a document in our collection. 

Author: ```suggestion
* Let's try to insert some data into our collection. 

You are used to seeing data in a tabular format due to the relational database structure. So let's start with a table as seen below and then convert it into a JSON object that can be inserted into our collection. 

|fname|lname|salary|departments|hiredate|
|---|---|---|---|---|
|john|doe|70000|sales, admin|2018-08-29|

Converting this table into an object, a set of `key:value` pairs surrounded by `{}` looks like this:
```

    ```
    {
	"fname": "john",
        "lname": "doe",
        "salary": 70000,
        "departments": ["sales", "admin"],
        "hiredate": "2018-08-29"
    }
    ```

> 	

Author: ```suggestion
Now that we have our JSON object, let's insert it into our collection. 
```

---

## Update Operations

The MongoDB query language supports 3 update operations:
* `updateOne()`
* `updateMany()`
* `replaceOne()` - [Read more](https://docs.mongodb.com/manual/reference/method/db.collection.replaceOne/#db.collection.replaceOne) :book: about this operator. 

**Exercise 15** :computer: 

Let's turn our attention towards movies made in 🇮🇳 ! I bet all of you have either heard of or watched "Taare Zameen Par" but unfortunately this collection hasn't got the name right! 

* Spend **5 minutes** :alarm_clock: to find out what the movie `title` is in our collection! Feel free to use the internet to look up identifying information like the release year or the actor/director/writer names to use in your query filter. 

* Once you have the name, let's update it to reflect the correct name "Taare Zameen Par". 

Here's how to perform the update operation:

> ```javascript

Reviewer: I :eyes: what you are trying to do here, you're show syntax, but you need to put that in an explanation somewhere above the query.  This query is not copy/paste ready because of the syntax demo.

db.movieDetails.updateOne(
    {<your uniquely identifying filter as key: value pair>}, 
    {"$set": {"title": "Taare Zameen Par"}}
    )
```

**Exercise 16** :computer: 

MongoDB offers multiple operators that can be used in its update operations. In the exercise above, we looked at the `$set` operator. 

|Name|Description|
|---|---|
|`$set`|Sets the value of a field in a document|
|`$unset`|Removes the specified field from a document|
|`$currentDate`|Sets the value of a field to current date, either as a Date or a Timestamp|
|`$inc`|Increments the value of a field by a specified amount|
|`$mul`|Multiplies the value of the field by the specified amount|
|`$min`|Only updates the field if the specified value is less than the existing field value.|
|`$max`|Only updates the field if the specified value is greater than the existing field value.|
|`$setOnInsert`|Sets the value of a field if an update results in an insert of a document. Has no effect on update operations that modify existing documents.|
|`$rename`|Renames a field|

> Take **10 minutes** :alarm_clock: to go through this list and write an interesting `$updateMany()` query. 

Reviewer: if you do this you may break the data for future exercises.  it may be better to have them make updates to the employee database that they created in the **Create** section of things.  Having them update the movie name was fine, but a complete overwrite of bulk data can be detrimental.  

Think about how this affects the other exercises and decide if you want to roll these 🎲 or not.  It's your call!

:arrow_right: Note that `updateMany()` will make the same modification to all the documents that match the given filter. 

---

## Delete Operations

The MongoDB query language supports 2 delete operations:
* `deleteOne()`
* `deleteMany()`

**Exercise 17** :computer: 

> In our `movieDetails` collection, there is only 1 movie from the `year` 2018 with a `title` "QQ Speed". To perform the exercise, delete this record. 

Reviewer: again, maybe we want this on a junk database.  Especially if any other exercise relies on counting or filtering on these parameters.  This is the **basic** exercise document, and I know you lean on the movies db heavily in the intermediate document.  

I also worry that you are stuck on this movies db because it's what you see in the videos.  Can we structure some activities around these labs to do the same things against other databases?

Reviewer: Maybe we teach against the movie database and all the activities and group work revolve around the other databases they have?  In my head this is a much better option.

Author: Added suggestions to where `updateMany()` and `deleteMany()` operations are performed against the dummy `employee_info` collection that was created earlier in this document.

**query:**
```javascript
db.movieDetails.deleteOne({"title": "QQ Speed"})
```
Alternatively, you could use:
```javascript
db.movieDetails.deleteOne({"year": 2018})
```
since this filter would also uniquely identify the movie "QQ Speed", considering this collection has only 1 movie from that year. 

**result:**
```javascript
{ "acknowledged" : true, "deletedCount" : 1 }
```

The result confirms that the `deleteOne()` operation was successful, at the same time telling us the number of documents that were deleted. 

**Exercise 18** :computer: 

Our `movieDetails` collection has a few movies from the 1800s!!! Spend **5 minutes** :alarm_clock: to delete all those movies using `deleteMany()`.

---

> ## Group Activities

Reviewer: why are we scrapping using other datasets that are already built in favor of them building a poor dataset with only three entries, one which they can't use later?

This is a good exercise for creating documents, and I'm all for it, but it should be meaningful and not arbitrary.  If you want to run an Activity like this, let's given them table data (`.csv` file) and have them convert it to JSON, and THEN run their inserts and such.

Author: Sure, that sounds good. You have a few `.csv` files across your repos and since you're more familiar with the data, which ones would you like to include in their group activity:

- [ ] [League of Legends - games.csv](https://github.com/mattdavis0351/mongodb/blob/master/datasets/league_of_legends/games.csv)
- [ ] [candy.csv](https://github.com/mattdavis0351/data/blob/master/candy.csv)
- [ ] [gender.csv](https://github.com/mattdavis0351/data/blob/master/gender.csv) 

I know that `mega city.csv` is being used for their Apache Pig training so I didn't include it in this list!

Reviewer: None, those all suck if I remember right.  I think at one point they were used with Talend, but not Mongo.  Nothing stops us from finding some interesting ones though, but these are a no-go

Reviewer: we're rolling with what we have!  You already know though

* Once you have the name, let's update it to reflect the correct name "Taare Zameen Par". 

Here's how to perform the update operation:

db.movieDetails.updateOne(
{<your uniquely identifying filter as `key: value` pair>}, {"$set": {"title": "Taare Zameen Par"}}
)

**Exercise 16** :computer: 

MongoDB offers multiple operators that can be used in its update operations. In the exercise above, we looked at the `$set` operator. 

|Name|Description|
|---|---|
|`$set`|Sets the value of a field in a document|
|`$unset`|Removes the specified field from a document|
|`$currentDate`|Sets the value of a field to current date, either as a Date or a Timestamp|
|`$inc`|Increments the value of a field by a specified amount|
|`$mul`|Multiplies the value of the field by the specified amount|
|`$min`|Only updates the field if the specified value is less than the existing field value.|
|`$max`|Only updates the field if the specified value is greater than the existing field value.|
|`$setOnInsert`|Sets the value of a field if an update results in an insert of a document. Has no effect on update operations that modify existing documents.|
|`$rename`|Renames a field|

> Take **10 minutes** :alarm_clock: to go through this list and write an interesting `$updateMany()` query. 

Author: ```suggestion
Take **10 minutes** :alarm_clock: to go through this list and write an interesting `$updateMany()` query against the `employee_info` collection we created earlier on. Ideally **Create**, **Update** and **Delete** operations would never be run against a production database. Hence for practise purposes, it's wise to perform this and the following exercises against a dummy collection like `employee_info`. 
```

Reviewer: I don't think it's fair to say that we wouldn't run those operations against a production database.  Here are a few scenarios where I need to update or add to a database:
- I created a new user account on your website
- I changed my email address
- I changed my email
- I asked you to close my account (but you needed to keep a record of my existence so you just toggled a flag making it seem like my account was deleted, but the truth is you just edited my ability to login)
- I changed my display picture

All of these are examples of mutations of data in a database that take place on a production deployment.

:arrow_right: Note that `updateMany()` will make the same modification to all the documents that match the given filter. 

---

## Delete Operations

The MongoDB query language supports 2 delete operations:
* `deleteOne()`
* `deleteMany()`

**Exercise 17** :computer: 

In our `movieDetails` collection, there is only 1 movie from the `year` 2018 with a `title` "QQ Speed". To perform the exercise, delete this record. 

**query:**
```javascript
db.movieDetails.deleteOne({"title": "QQ Speed"})
```
Alternatively, you could use:
```javascript
db.movieDetails.deleteOne({"year": 2018})
```
since this filter would also uniquely identify the movie "QQ Speed", considering this collection has only 1 movie from that year. 

**result:**
```javascript
{ "acknowledged" : true, "deletedCount" : 1 }
```

The result confirms that the `deleteOne()` operation was successful, at the same time telling us the number of documents that were deleted. 

**Exercise 18** :computer: 

> Our `movieDetails` collection has a few movies from the 1800s!!! Spend **5 minutes** :alarm_clock: to delete all those movies using `deleteMany()`.

Author: ```suggestion
Time to turn our attention back to the `employee_info` collection! Spend **5 minutes** :alarm_clock: to delete employees using `deleteMany()`.
```

Reviewer: what about two **delete()** exercises?

- delete({department:<some department with more than one person>})
  - this would show that the "first" doc get manipulated and thus highlights unintended consequences to an operation like this.
- deleteMany({departments: "managers"})
  - this should remove all documents that belong to managers

Note: those queries are pseudo code and need to be formatted correctly.
